File: scripts/solar/solar-wc3/lib/simulate/ItemSimulate.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 9,["11"] = 9,["12"] = 9,["14"] = 32,["15"] = 33,["18"] = 36,["19"] = 37,["22"] = 40,["23"] = 31,["24"] = 48,["25"] = 49,["26"] = 51,["27"] = 52,["29"] = 55,["30"] = 56,["31"] = 57,["33"] = 60,["34"] = 61,["35"] = 62,["36"] = 62,["37"] = 62,["38"] = 62,["39"] = 62,["40"] = 62,["41"] = 62,["42"] = 63,["44"] = 66,["45"] = 67,["46"] = 68,["47"] = 69,["49"] = 72,["50"] = 73,["51"] = 74,["53"] = 75,["54"] = 77,["55"] = 78,["56"] = 79,["57"] = 80,["58"] = 82,["59"] = 84,["61"] = 87,["62"] = 48,["63"] = 95,["64"] = 97,["65"] = 98,["66"] = 99,["68"] = 101,["69"] = 95,["70"] = 108,["71"] = 109,["74"] = 112,["75"] = 114,["76"] = 115,["77"] = 116,["78"] = 117,["79"] = 115,["80"] = 120,["81"] = 121,["82"] = 122,["83"] = 123,["84"] = 121,["85"] = 126,["86"] = 127,["87"] = 128,["88"] = 129,["89"] = 127,["90"] = 134,["91"] = 135,["92"] = 136,["93"] = 137,["94"] = 138,["95"] = 135,["96"] = 108,["97"] = 11,["98"] = 14,["99"] = 14,["100"] = 14,["101"] = 14,["102"] = 14,["103"] = 14,["104"] = 14,["105"] = 14,["106"] = 14,["107"] = 14,["108"] = 14,["109"] = 14,["110"] = 14,["111"] = 14,["112"] = 14,["113"] = 14,["114"] = 14,["115"] = 14,["116"] = 14,["117"] = 14,["118"] = 14,["119"] = 14,["120"] = 14,["121"] = 14,["122"] = 14,["123"] = 14,["124"] = 14,["125"] = 14,["126"] = 14,["127"] = 14,["128"] = 14,["129"] = 14,["130"] = 14,["131"] = 14,["132"] = 14,["133"] = 22,["134"] = 26,["135"] = 28,["136"] = 43});
local ____exports = {}
local ____SingletonUtil = require("solar.solar-common.util.lang.SingletonUtil")
local SingletonUtil = ____SingletonUtil.default
local ____LangUtil = require("solar.solar-common.util.lang.LangUtil")
local LangUtil = ____LangUtil.default
____exports.default = __TS__Class()
local ItemSimulate = ____exports.default
ItemSimulate.name = "ItemSimulate"
function ItemSimulate.prototype.____constructor(self)
    if SingletonUtil:notFirstTime(____exports.default) then
        print("不能重复new ItemSimulate()")
        return
    end
    if not ____exports.default.finder then
        print("请先配置物品模拟信息查找器")
        return
    end
    ____exports.default:_sl_init()
end
function ItemSimulate.getRealObjId(self, itemid)
    local intId = LangUtil:getIntId(itemid)
    if _g_objs.item[id2string(intId)] then
        return intId
    end
    local templateId = ____exports.default.ft_itemIdMap[intId]
    if templateId then
        return templateId
    end
    if ____exports.default.index >= #____exports.default.templateIds then
        log.errorWithTraceBack("模拟物品已达到上限:" .. tostring(____exports.default.index))
        DisplayTimedTextToPlayer(
            GetLocalPlayer(),
            0,
            0,
            30,
            "模拟物品已达到上限(请重开游戏，减少物品类型出现总数):" .. tostring(____exports.default.index)
        )
        return 0
    end
    local simulateData = ____exports.default:finder(intId)
    if simulateData == nil then
        log.errorWithTraceBack((("你正在创建一个不存在的物品:" .. id2string(intId)) .. " -> ") .. tostring(intId))
        return 0
    end
    templateId = FourCC(____exports.default.templateIds[____exports.default.index + 1])
    ____exports.default.ft_itemIdMap[intId] = templateId
    ____exports.default.tf_itemIdMap[templateId] = intId
    local ____exports_default_0, ____index_1 = ____exports.default, "index"
    ____exports_default_0[____index_1] = ____exports_default_0[____index_1] + 1
    EXSetItemDataString(templateId, 1, simulateData.Art)
    EXSetItemDataString(templateId, 4, simulateData.Tip)
    EXSetItemDataString(templateId, 3, simulateData.Ubertip)
    if isEmbedJapi then
        EXSetItemDataString(templateId, 2, simulateData.Tip)
        EXSetItemDataString(templateId, 5, simulateData.Ubertip)
    end
    return templateId
end
function ItemSimulate.getLogicId(self, intId)
    local logicId = ____exports.default.tf_itemIdMap[intId]
    if logicId then
        return logicId
    end
    return intId
end
function ItemSimulate._sl_init(self)
    if ____exports.default._sl_isInit then
        return
    end
    ____exports.default._sl_isInit = true
    local oldCreateItem = CreateItem
    _G.CreateItem = function(itemid, x, y)
        local realObjId = ____exports.default:getRealObjId(itemid)
        return oldCreateItem(realObjId, x, y)
    end
    local oldUnitAddItemById = UnitAddItemById
    _G.UnitAddItemById = function(whichUnit, itemid)
        local realObjId = ____exports.default:getRealObjId(itemid)
        return oldUnitAddItemById(whichUnit, realObjId)
    end
    local oldUnitAddItemToSlotById = UnitAddItemToSlotById
    _G.UnitAddItemToSlotById = function(whichUnit, itemid, itemSlot)
        local realObjId = ____exports.default:getRealObjId(itemid)
        return oldUnitAddItemToSlotById(whichUnit, realObjId, itemSlot)
    end
    local oldGetItemTypeId = GetItemTypeId
    _G.GetItemTypeId = function(i)
        local realItemTypeId = oldGetItemTypeId(i)
        local logicId = ____exports.default:getLogicId(realItemTypeId)
        return logicId
    end
end
ItemSimulate.finder = nil
ItemSimulate.templateIds = {
    "azhr",
    "bzbe",
    "bzbf",
    "ches",
    "dkfw",
    "dphe",
    "dthb",
    "engs",
    "glsk",
    "gmfr",
    "gopr",
    "jpnt",
    "k3m1",
    "k3m2",
    "k3m3",
    "ktrm",
    "kybl",
    "kygh",
    "kymn",
    "kysn",
    "ledg",
    "mgtk",
    "mort",
    "phlt",
    "sclp",
    "sehr",
    "shwd",
    "skrt",
    "sorf",
    "soul",
    "thle",
    "wolg",
    "wtlg"
}
ItemSimulate.ft_itemIdMap = {}
ItemSimulate.tf_itemIdMap = {}
ItemSimulate.index = 0
ItemSimulate._sl_isInit = false
return ____exports

File: src/solar/solar-wc3/lib/simulate/ItemSimulate.ts
import SingletonUtil from "@/SingletonUtil";
import LangUtil from "@/LangUtil";

/**
 * 物品模拟
 * 可模拟无主动使用的装备
 * （应用场景：ORPG地图上万装备时 建实际物编太卡 单局同时出现的物品不会超过太多 可以用1000个模板物品来运转这上万个装备）
 */
export default class ItemSimulate {
    //配置一个查找器 根据需要模拟的物品id(4个字符的256进制) 获得需要模拟的物品信息
    static finder: (itemId: number) => iobj = null;

    //空技能的 模板物品id 可手动指定更多模板id 以免出现模拟物品使用不够的情况
    static templateIds = ["azhr", "bzbe", "bzbf", "ches", "dkfw", "dphe", "dthb", "engs", "glsk", "gmfr", "gopr", "jpnt",
        "k3m1", "k3m2", "k3m3", "ktrm", "kybl", "kygh", "kymn", "kysn", "ledg", "mgtk", "mort", "phlt", "sclp", "sehr",
        "shwd", "skrt", "sorf", "soul", "thle", "wolg", "wtlg"]
    /**
     * 整数物品的id  与模板id的 映射关系
     */
    static ft_itemIdMap: {
        [fakeId: number]: number
    } = {}
    //模板id 与模拟物品id 的映射
    static tf_itemIdMap: {
        [templateId: number]: number
    } = {}
    //使用数量 这个数量超过模板物品id 会报错 请保证模板id 足够当前游戏局使用
    static index: number = 0;


    constructor() {
        if (SingletonUtil.notFirstTime(ItemSimulate)) {
            print("不能重复new ItemSimulate()")
            return;
        }
        if (!ItemSimulate.finder) {
            print("请先配置物品模拟信息查找器")
            return;
        }
        ItemSimulate._sl_init();
    }

    private static _sl_isInit = false;

    /**
     * 获取真实的物编id
     */
    static getRealObjId(itemid: string | number): number {
        let intId = LangUtil.getIntId(itemid);
        //判断是否已是真实id
        if (_g_objs.item[id2string(intId)]) {
            return intId;
        }
        //从缓存中拿
        let templateId = ItemSimulate.ft_itemIdMap[intId];
        if (templateId) {
            return templateId;
        }
        //模拟id 从模板id池 申请一个id与此id对应
        if (ItemSimulate.index >= ItemSimulate.templateIds.length) {
            log.errorWithTraceBack("模拟物品已达到上限:" + ItemSimulate.index)
            DisplayTimedTextToPlayer(GetLocalPlayer(), 0, 0, 30, "模拟物品已达到上限(请重开游戏，减少物品类型出现总数):" + ItemSimulate.index);
            return 0;
        }
        //判断是否是模拟物品
        let simulateData = ItemSimulate.finder(intId);
        if (simulateData == null) {
            log.errorWithTraceBack("你正在创建一个不存在的物品:" + id2string(intId) + " -> " + intId)
            return 0;
        }
        //
        templateId = FourCC(ItemSimulate.templateIds[ItemSimulate.index])
        ItemSimulate.ft_itemIdMap[intId] = templateId
        ItemSimulate.tf_itemIdMap[templateId] = intId
        ItemSimulate.index++;
        //模拟数据 套到 模板物品上
        EXSetItemDataString(templateId, 1, simulateData.Art)
        EXSetItemDataString(templateId, 4, simulateData.Tip)
        EXSetItemDataString(templateId, 3, simulateData.Ubertip)
        if (isEmbedJapi) {
            //设置名字
            EXSetItemDataString(templateId, 2, simulateData.Tip)
            //设置物品描述
            EXSetItemDataString(templateId, 5, simulateData.Ubertip)
        }
        //
        return templateId;
    }

    /**
     * 获取逻辑物品id
     * (如果是模板id 则返回模拟的物品id)
     * @param intId
     */
    static getLogicId(intId: number): number {
        //从缓存中拿
        let logicId = ItemSimulate.tf_itemIdMap[intId];
        if (logicId) {
            return logicId;
        }
        return intId;
    }

    /**
     * hook 原生函数
     * @private
     */
    private static _sl_init() {
        if (ItemSimulate._sl_isInit) {
            return
        }
        ItemSimulate._sl_isInit = true;
        //one case
        let oldCreateItem = CreateItem;
        _G.CreateItem = function (this: void, itemid: number | string, x: number, y: number): item {
            let realObjId = ItemSimulate.getRealObjId(itemid);
            return oldCreateItem(realObjId, x, y);
        }
        //one case
        let oldUnitAddItemById = UnitAddItemById;
        _G.UnitAddItemById = function (this: void, whichUnit: unit, itemid: number | string): item {
            let realObjId = ItemSimulate.getRealObjId(itemid);
            return oldUnitAddItemById(whichUnit, realObjId);
        }
        //one case
        let oldUnitAddItemToSlotById = UnitAddItemToSlotById;
        _G.UnitAddItemToSlotById = function (this: void, whichUnit: unit, itemid: number | string, itemSlot: number): boolean {
            let realObjId = ItemSimulate.getRealObjId(itemid);
            return oldUnitAddItemToSlotById(whichUnit, realObjId, itemSlot);
        }

        //使用
        //one case
        let oldGetItemTypeId = GetItemTypeId;
        _G.GetItemTypeId = function (this: void, i: item): number {
            let realItemTypeId = oldGetItemTypeId(i);
            let logicId = ItemSimulate.getLogicId(realItemTypeId);
            return logicId;
        }
    }


}